File: claude-code-sdk/src/main/kotlin/com/claudecodeplus/sdk/types/McpTypes.kt
package com.claudecodeplus.sdk.types

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

/**
 * MCP tool definition based on Python SDK SdkMcpTool.
 */
@Serializable
data class SdkMcpTool(
    val name: String,
    val description: String,
    val inputSchema: JsonElement, // JSON schema for input validation
    // Note: handler is not serializable, handled separately in runtime
)

/**
 * MCP server instance configuration.
 */
@Serializable
data class McpServerInstance(
    val name: String,
    val config: McpServerConfig,
    val tools: List<SdkMcpTool> = emptyList(),
    val status: McpServerStatus = McpServerStatus.DISCONNECTED
)

/**
 * MCP server connection status.
 */
enum class McpServerStatus {
    CONNECTED,
    DISCONNECTED,
    CONNECTING,
    ERROR
}

/**
 * MCP resource information.
 */
@Serializable
data class McpResource(
    val uri: String,
    val name: String,
    val description: String? = null,
    val mimeType: String? = null
)

/**
 * MCP tool execution request.
 */
@Serializable
data class McpToolRequest(
    val serverName: String,
    val toolName: String,
    val arguments: JsonElement
)

/**
 * MCP tool execution response.
 */
@Serializable
data class McpToolResponse(
    val success: Boolean,
    val result: JsonElement? = null,
    val error: String? = null,
    val metadata: Map<String, JsonElement> = emptyMap()
)

/**
 * MCP prompt template.
 */
@Serializable
data class McpPrompt(
    val name: String,
    val description: String? = null,
    val arguments: List<McpPromptArgument> = emptyList()
)

/**
 * MCP prompt argument definition.
 */
@Serializable
data class McpPromptArgument(
    val name: String,
    val description: String? = null,
    val required: Boolean = true
)

/**
 * MCP server capabilities.
 */
@Serializable
data class McpServerCapabilities(
    val logging: McpLoggingCapability? = null,
    val prompts: McpPromptsCapability? = null,
    val resources: McpResourcesCapability? = null,
    val tools: McpToolsCapability? = null
)

/**
 * MCP logging capability.
 */
@Serializable
data class McpLoggingCapability(
    val enabled: Boolean = false
)

/**
 * MCP prompts capability.
 */
@Serializable
data class McpPromptsCapability(
    val listChanged: Boolean = false
)

/**
 * MCP resources capability.
 */
@Serializable
data class McpResourcesCapability(
    val subscribe: Boolean = false,
    val listChanged: Boolean = false
)

/**
 * MCP tools capability.
 */
@Serializable
data class McpToolsCapability(
    val listChanged: Boolean = false
)

/**
 * MCP client capabilities.
 */
@Serializable
data class McpClientCapabilities(
    val experimental: Map<String, JsonElement> = emptyMap(),
    val sampling: Map<String, JsonElement> = emptyMap()
)

/**
 * MCP initialization result.
 */
@Serializable
data class McpInitializeResult(
    val protocolVersion: String,
    val capabilities: McpServerCapabilities,
    val serverInfo: McpServerInfo
)

/**
 * MCP server information.
 */
@Serializable
data class McpServerInfo(
    val name: String,
    val version: String
)

/**
 * MCP client information.
 */
@Serializable
data class McpClientInfo(
    val name: String = "claude-code-plus",
    val version: String = "1.0.0"
)

File: vscode-extension/src/sdk/claude/types/mcpTypes.ts
/**
 * MCP (Model Context Protocol) types for Claude Agent SDK
 * Translated from Kotlin: claude-agent-sdk/src/main/kotlin/com/asakii/claude/agent/sdk/types/McpTypes.kt
 */

import type { JsonValue, JsonObject } from './common';
import type { McpServerConfig } from './options';

/**
 * MCP tool definition based on Python SDK SdkMcpTool.
 */
export interface SdkMcpTool {
  name: string;
  description: string;
  /** JSON schema for input validation */
  inputSchema: JsonValue;
  // Note: handler is not serializable, handled separately in runtime
}

/**
 * MCP server connection status.
 */
export type McpServerStatus = 'CONNECTED' | 'DISCONNECTED' | 'CONNECTING' | 'ERROR';

/**
 * MCP server status enum values.
 */
export const McpServerStatuses = {
  CONNECTED: 'CONNECTED',
  DISCONNECTED: 'DISCONNECTED',
  CONNECTING: 'CONNECTING',
  ERROR: 'ERROR',
} as const;

/**
 * MCP server instance configuration.
 */
export interface McpServerInstance {
  name: string;
  config: McpServerConfig;
  tools?: SdkMcpTool[];
  status?: McpServerStatus;
}

/**
 * MCP resource information.
 */
export interface McpResource {
  uri: string;
  name: string;
  description?: string;
  mimeType?: string;
}

/**
 * MCP tool execution request.
 */
export interface McpToolRequest {
  serverName: string;
  toolName: string;
  arguments: JsonValue;
}

/**
 * MCP tool execution response.
 */
export interface McpToolResponse {
  success: boolean;
  result?: JsonValue;
  error?: string;
  metadata?: Record<string, JsonValue>;
}

/**
 * MCP prompt template.
 */
export interface McpPrompt {
  name: string;
  description?: string;
  arguments?: McpPromptArgument[];
}

/**
 * MCP prompt argument definition.
 */
export interface McpPromptArgument {
  name: string;
  description?: string;
  required?: boolean;
}

/**
 * MCP server capabilities.
 */
export interface McpServerCapabilities {
  logging?: McpLoggingCapability;
  prompts?: McpPromptsCapability;
  resources?: McpResourcesCapability;
  tools?: McpToolsCapability;
}

/**
 * MCP logging capability.
 */
export interface McpLoggingCapability {
  enabled?: boolean;
}

/**
 * MCP prompts capability.
 */
export interface McpPromptsCapability {
  listChanged?: boolean;
}

/**
 * MCP resources capability.
 */
export interface McpResourcesCapability {
  subscribe?: boolean;
  listChanged?: boolean;
}

/**
 * MCP tools capability.
 */
export interface McpToolsCapability {
  listChanged?: boolean;
}

/**
 * MCP client capabilities.
 */
export interface McpClientCapabilities {
  experimental?: Record<string, JsonValue>;
  sampling?: Record<string, JsonValue>;
}

/**
 * MCP initialization result.
 */
export interface McpInitializeResult {
  protocolVersion: string;
  capabilities: McpServerCapabilities;
  serverInfo: McpServerInfo;
}

/**
 * MCP server information.
 */
export interface McpServerInfo {
  name: string;
  version: string;
}

/**
 * MCP client information.
 */
export interface McpClientInfo {
  name?: string;
  version?: string;
}

/**
 * Default MCP client info.
 */
export const defaultMcpClientInfo: McpClientInfo = {
  name: 'claude-code-plus',
  version: '1.0.0',
};
